### PR TITLE
Random net

### PR DESCRIPTION
**Pull request by thyamu**

### CODE REVIEW

File: docs/source/boolean.rst [hunk 1]
```diff
@@ -34,12 +34,18 @@ Yeast Networks
...
 
     .. autoattribute:: neet.boolean.examples.p53_dmg
-        :annotation: = <neet.boolean.WTNetwork object>        
+        :annotation: = <neet.boolean.WTNetwork object>
+
+
+Random networks
+^^^^^^^^^^^^^^^
+
+.. automodule:: neet.boolean.rannetwork
```
Comment (hbsmith) [inline docs/source/boolean.rst]: Not sure, but this line might need to be indented?

File: neet/boolean/rannetwork.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
+# Copyright 2017 ELIFE. All rights reserved.
+# Use of this source code is governed by a MIT
+# license that can be found in the LICENSE file.
+
+import sys
+import os
+import random as ran
+import networkx as nx
```
Comment (siyuzhou) [inline neet/boolean/rannetwork.py]: `sys`,`os` and `networkx` are not used.

File: neet/boolean/rannetwork.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            for k in range(length_arr):
+                if edges_swiped:
+                    break
+                for l in range(length_arr):
+                    if edges_swiped:
+                        break
+                    if k == l: # to perserve self-loops
+                        continue
+                    if arr[i, j] != arr[k, l]: # edge-swipe is allowed only between two edges sharing the same weight
+                        continue
```
Comment (siyuzhou) [inline neet/boolean/rannetwork.py]: I don't think it is wrong, but I imagine many networks to have continuously valued edges. Limiting edge exchange between those of the same weight may not serve future need. But we can start another issue.

File: neet/boolean/rannetwork.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                continue
+            if i == j:   # to preserve self-loops
+                continue
+            edges_swiped = False
+            for k in range(length_arr):
+                if edges_swiped:
+                    break
+                for l in range(length_arr):
+                    if edges_swiped:
+                        break
```
Comment (bcdaniels) [inline neet/boolean/rannetwork.py]: Maybe I'm confused, but does this logic mean that there is a bias to use nodes with smaller index more frequently in swaps?  

It looks like for each (i,j) we select one (k,l) to swap with, going through the possible (k,l) pairs one by one and picking that pair with probability 0.5.  But we go through the pairs always in the same order, starting with k=0, l=0.  For a large network, would that mean that nodes with large indices almost never get picked as k and l?

File: docs/source/boolean.rst (hunk 1 above)
Comment (thyamu) [inline docs/source/boolean.rst]: Done. I am not sure if that is necessary but I made the indentation in the document consistently.

File: neet/boolean/rannetwork.py (hunk 2 above)
Comment (thyamu) [inline neet/boolean/rannetwork.py]: Done. Thanks!

File: neet/boolean/rannetwork.py (hunk 4 above)
Comment (thyamu) [inline neet/boolean/rannetwork.py]: @bcdaniels You are absolutely right. I was thinking of the problem but concluded it won't affect much for small networks that we already had. However, now we will have bigger networks, I think I will change the part accordingly.

File: neet/boolean/rannetwork.py (hunk 3 above)
Comment (thyamu) [inline neet/boolean/rannetwork.py]: @siyuzhou Yes, I chose the rewiring under the very strict condition, which won't work for networks with continuous weights. Please go ahead and issue separate issue for it.

File: neet/boolean/rannetwork.py (hunk 4 above)
Comment (siyuzhou) [inline neet/boolean/rannetwork.py]: I agree with @bcdaniels, and I think even for a small network the bias could be noticeable. I suggest we still merge it so every one can get along with the workflow. In the mean time, @thyamu can you work on an unbiased algorithm and update the file?

**Comment by thyamu** [comment]
@siyuzhou, @bcdaniels and @dglmoore I updated rannetwork for an unbaised selection of edges for rewiring. Please check them before merging the pull request. Thanks!

File: neet/boolean/rannetwork.py (hunk 4 above)
Comment (bcdaniels) [inline neet/boolean/rannetwork.py]: Nice—better.  Though now I have to be a stickler and say that I think this slightly biases having l = (k+1)%length_arr.  That is: initially having any pair (k, l) is equally likely, but if we pick k=l we add one to l and use that, which makes that case slightly more likely.

**Comment by bcdaniels** [comment]
I made another comment on the updated code.  Also, generally I would recommend having a way to set the random seed (for example, at the moment the tests run depend on the particular random instance, and it can get very confusing if tests sometimes pass and sometimes fail).

File: neet/boolean/rannetwork.py (hunk 4 above)
Comment (thyamu) [inline neet/boolean/rannetwork.py]: good point! That can be easily fixed by chaning r = np.random.choice(length_arr, 2, replace=True) to  r = np.random.choice(length_arr, 2, replace=False).

Comment (bcdaniels) [inline neet/boolean/rannetwork.py]: True!

**Comment by dglmoore** [comment]
@bcdaniels You can set the seed before entering the function using `numpy.random.seed`. That being said, adding a `seed` argument to all random functions might be advisable.

Regarding the unit tests, we need these functions to generate pseudo-random results. As you pointed out, that could make finding and fixing bugs a nightmare because each time you run the tests you'll get a different result. One solution is to choose a random seed, and print the seed when a test fails. All of the assertion methods accept a `msg` argument:

```python
def test_preserving_degrees(self):
    # generate a random 32-bit seed
    seed = np.random.randint(0, 2147483647)
    # seed the random number generator
    np.random.seed(seed)

    weights = np.random.randint(low = -1, high = 2, size = (10,10)) 
    net = WTNetwork(copy.copy(weights)) 
    ran_net = rnet.rewiring_fixed_degree(net) 

    G = nx.from_numpy_matrix(net.weights, create_using = nx.DiGraph()) 
    ranG = nx.from_numpy_matrix(ran_net.weights, create_using = nx.DiGraph()) 

    InDegree = list(G.in_degree(weight = 'weight').values()) 
    ranInDegree = list(ranG.in_degree(weight = 'weight').values()) 

    OutDegree = list(G.out_degree(weight = 'weight').values()) 
    ranOutDegree = list(ranG.out_degree(weight = 'weight').values()) 

    # add a message to each assertion so we know for which seed the tests fail
    self.assertEqual(InDegree, ranInDegree, msg="random seed: {}".format(seed)) 
    self.assertEqual(OutDegree, ranOutDegree, msg="random seed: {}".format(seed))
```

@thyamu, @hbsmith @bcdaniels Are we happy with this pull request and ready to merge it, or do we want to do a little more work on it?

**Comment by bcdaniels** [comment]
My suggestions:

1) Change the logic of the rannetwork functions so that each pair (k,l) that meets the conditions to be swappable has equal probability of being chosen.  In rewiring_fixed_degree, I think this will require more than changing to replace=False.

2) Personally, I want unit tests to be strictly deterministic—so I would hard-code a seed in the unit test.  (If you want to test robustness to changing the seed, then I would have the unit test explicitly check a few seeds.)  I suppose this is a matter of taste.  I prefer to have complete control over my randomness.  :)

**Comment by dglmoore** [comment]
> 2. Personally, I want unit tests to be strictly deterministic—so I would hard-code a seed in the unit test. (If you want to test robustness to changing the seed, then I would have the unit test explicitly check a few seeds.) I suppose this is a matter of taste. I prefer to have complete control over my randomness. :)

I don't think that is unreasonable. @thyamu, I think the best way to do this is set the seed within the `unittest.TestCase.setUp` method. This method is called before each test is run. So, for example, you might have something like

```python
class TestRandom(unittest.TestCase):

    def setUp(self):
        """
        Setup the unit tests by setting a fixed seed
        """
        np.random.seed(1234)
```

**Comment by dglmoore** [comment]
@thyamu, @hbsmith and @bcdaniels, where are we with this one?

**Comment by bcdaniels** [comment]
It would be nice to have some random networks to compare for our results extravaganza on Tuesday.  Is this pull request close to being ready?

**Comment by thyamu** [comment]
@bcdaniels This is not ready yet. 

Firstly, I am considering to modify the part to choose a pair of edges to swipe from the set of all possible edges instead of going through the matrix, as you suggested. I agree that in the way we can remove the bias in selecting edges. However, I am not sure how much that bias will affect the resulting networks and how the modification will affect the performance for the case of large networks. 

Secondly, the code is written for WTnetwork. I need to modify it so that we can use it for logic networks.

**Pull request closed without merging**